Upgrade to Rails 7

stefannibrasil opened the pull request:
This is one of the dependencies we need to upgrade to Rails 7 first before upgrading the main CK's app.

The major changes are:

- removal of deprecated `active_record.legacy_connection_handling`. This was deprecated in 7.0 and removed in 7.1. This config no longer has effects since multiple databases config in Rails has to be configured manually. It is now set to false by default on new rails apps.
- `fixture_path` has been renamed to `fixture_paths` because it's an array that accepts multiples paths.

The other file changes are made so the gem is running on Rails 7.1.2.

sbs-cats committed:
Upgrade Rails to 7.1.2

sbs-cats committed:
Update config.load_defaults

From Rails guides:
https://guides.rubyonrails.org/configuring.html#versioned-default-values

config.load_defaults loads default configuration values for a target
version and all versions prior. With this commit, config.load_defaults 7.1
will load defaults for all versions up to and including version 7.1.

sbs-cats committed:
Remove deprecated `legacy_connection_handling` setter

This was deprecated in 7.0 and removed in 7.1. This config no longer has
effects since multiple databases config in Rails has to be configured
manually. It is now set to false by default.

sbs-cats committed:
Fix fixtures_path deprecation

With Rails 7.1.2, configuring `fixture_path` has been deprecated in
favor of using an aray of `fixture_paths`.

sbs-cats committed:
ActiveRecord::Base is not accessible anymore

since https://github.com/rails/rails/pull/42445, this method has been
changed from cattr_accessor to attr_accessor in ActiveRecord.

sbs-cats committed:
Allow Rails > 6.1 and < 7 versions

otherwise, we can't upgrade this gem on the main app ;)

keegnotrub commented on impressionist.gemspec:
```diff
@@ -20,7 +20,7 @@ Gem::Specification.new do |s|
   s.add_dependency "friendly_id"
   s.add_dependency 'nokogiri', RUBY_VERSION < '2.1.0' ? '~> 1.6.0' : '~> 1'
-  s.add_dependency 'rails', '>= 3.2.15', '< 7.1'
+  s.add_dependency 'rails', '>= 6.1.7.6', '< 7.1.2'
```
If these changes allow us to support Rails 7.1, I wonder if we can just remove the second constraint completely?

```ruby
  s.add_dependency 'rails', '>= 6.1.7.6'
```

stefannibrasil replied:
yep, great call, thank you!

changed it in [a69b466](https://github.com/ConvertKit/impressionist/pull/3/commits/a69b466e7df135abeddb9e387572ff4d2a607891)

sbs-cats committed:
Support any Rails version >= 6.1..7.6

These changes allow us to support Rails 7.1, so we can remove the
second constraint completely. Otherwise, when we upgrade it to > 7.1,
we will need to update the version again.

keegnotrub commented on lib/generators/active_record/impressionist_generator.rb:
```diff
@@ -7,7 +7,7 @@ class ImpressionistGenerator < Rails::Generators::Base
       def self.next_migration_number(dirname)
         sleep 1
-        if ActiveRecord::Base.timestamped_migrations
+        if ActiveRecord.timestamped_migrations
```
I'm not 100% if this change is supported in Rails 6.1?  Either way I don't think we use this generator, but in case its important:

```
> ActiveRecord.timestamped_migrations
(irb):1:in `<main>': undefined method `timestamped_migrations' for ActiveRecord:Module (NoMethodError)
 > ActiveRecord::Base.timestamped_migrations
=> true
```

stefannibrasil replied:
Yeah, considering that we only use it for the landing pages and not for anything else, I wouldn't worry about the generators.  This gem is probably gonna be removed right after the upgrade work is done anyway, so I am gonna leave as it is 👍  thank you for raising that, though 💯 

We have this in the main app so this file is probably used only when the generator is run for the first time:

https://github.com/ConvertKit/convertkit/blob/main/db/migrate/20130114172639_create_impressions_table.rb

keegnotrub commented on spec/spec_helper.rb:
```diff
@@ -23,7 +23,7 @@
 # See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
 RSpec.configure do |config|
-  config.fixture_path = "spec/fixtures"
+  config.fixture_paths = ["spec/fixtures"]
```
Similarly, I _believe_ this may only be supported in newer versions of Rails
```
> require "rspec/rails"
=> true
> RSpec.configure { |c| c.fixture_paths = ["spec/fixtures"] }
NoMethodError: undefined method `fixture_paths=' for #<RSpec::Core::Configuration:0x00007f9cfc9968d0
Did you mean?  fixture_path=
               fixture_path
               fixture_path?
from (pry):8:in `block in __pry__'
> RSpec.configure { |c| c.fixture_path = "spec/fixtures" }
=> "spec/fixtures"

```

stefannibrasil replied:
Yeah, good call! If this was being used for other things and outside of convertkit, I would take more care of these cases but going with the comment above, let's go with this for now.

keegnotrub reviewed (approved):
Good to go, for what its worth, I imagine these changes are more a blocker for Rails 7.1 vs Rails 7.0 correct?  If so, may want to prioritize changing the git SHA ahead of the Rails 7.1 upgrade instead of the Rails 7.0?  Although perhaps this helps to silence some warnings? Either way looks good, thank you!

stefannibrasil commented:
Thanks, @keegnotrub !

> If so, may want to prioritize changing the git SHA ahead of the Rails 7.1 upgrade instead of the Rails 7.0?

I've read this a couple of times and still can't follow, sorry 🙈 could you describe it a bit more, please?

keegnotrub commented:
Apologies, just wondering if we'd necessarily need this updated in order to get to Rails 7.0 (as opposed to Rails 7.1) for the main ckapp repo.

stefannibrasil commented:
Oh, I see what you mean, now, thanks! It was more to have one less thing to be done later than anything else. Also, I tested the changes in staging and the visitors are working, which is the only thing, AFAIK, that we rely on. So if it works, great, right?! 😆
